Implement new questions replacement

AlexVelezLl opened the pull request:
## Summary
* Implement new manual questions replacement according to the [specs](https://www.figma.com/design/lVJt5ukOrxS9qay0rXy7GC/0.18-Coach-updates?node-id=2437-42537&t=qukwYd3oF7LGiIpA-0)
* Remove old question replacement component, and related methods.


https://github.com/user-attachments/assets/abb7af9b-e00c-4483-8c66-f75b716e9ed8



## References
Closes part of #13108, auto-replace is out of scope of this PR

## Reviewer guidance
* Go to coach/quizzes/create new quiz
* Add questions to the section either manually or randomly.
* Use the new "replace" button in the question accodion and replace the question with one/multiple questions in manual/random selection mode.
* Try replacing questions from different parts of the questions list and check that it works well in all cases

AlexVelezLl committed:
Implement new replace question behaviour

AlexVelezLl committed:
Remove old questions replacement component

marcellamaki commented:
@pcenov and @radinamatic this will be a top priority for testing tomorrow - i'll be starting code review now

AlexVelezLl committed:
Remove duplicated method call

nucleogenesis commented on kolibri/plugins/coach/assets/src/composables/useQuizCreation.js:
```diff
@@ -115,7 +126,24 @@ export default function useQuizCreation() {
   }
 
-  function addQuestionsToSectionFromResources({ sectionIndex, resourcePool, questionCount }) {
+  function _insertItemsAt(baseArray, itemsToInsert, index) {
```
Did you consider using [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice#syntax) instead of making a function? For whatever is passed in as `baseArray` you could do `baseArray.splice(index, 0, itemsToInsert)` - although it does mutate the array in place.

Not blocking as this looks good overall and I see you're doing checks on the passed index as well, it just got me thinking

AlexVelezLl replied:
Oh I didnt :sweat_smile:. I just refactored the function to use the `splice` making a copy of the `baseArray` instead and it looks so much cleaner. Thanks @nucleogenesis!

nucleogenesis reviewed (commented):
Didn't see anything blocking - just added a question. Overall the code LGTM, but will leave approval to the assigned reviewers

AlexVelezLl committed:
Refactor insertItemsAt to use splice

pcenov commented:
1. After having added questions to a section if I select some or all of the questions the "Select all", "Replace" and "Delete" options are not working. Looking at Figma it seems that the "Replace" option should be removed but the other 2 should be functioning normally unless otherwise specified:

https://github.com/user-attachments/assets/b311b168-fe83-4f3f-9b2e-2ce8397969e2

2. When I select a single question now it is possible to replace it with multiple other questions if there are available questions but at the same time I can't see a way of bulk replacing several of my questions - mentioning this as confusing and not very comfortable or intuitive.
3. In addition to the above point, after I have just added several questions as a replacement to the one I've selected, then the question list is not refreshed automatically and there's no snackbar message confirming the successful replacement:

https://github.com/user-attachments/assets/8ee314c1-0dee-4fab-b13a-f89429b3dec6

4. When replacing questions I would expect the label of the button "Add N questions" to be changed to "Replace N questions". This is also valid for the title of the modal which is confusingly also labeled "Add questions to "Section 1". Lastly if I have already added the max number of questions when I try to replace 1 I am seeing the button "Add N questions" grayed out and labeled "Add 0 questions" while I would expect it to be "Replace 1 question":

https://github.com/user-attachments/assets/dd174d47-3469-4c96-ac5d-3fd13229a6f5

5. If I switch off the manual selection it's not clear at all why I can't add additional questions:

https://github.com/user-attachments/assets/d5992b98-2f9e-466d-a544-0f3042f3b190

AlexVelezLl committed:
Remove old questions replacement button

AlexVelezLl commented:
Ah my apologies, I somehow messed things up with the version control and it seems that some changes were reverted :sweat_smile:. Just pushed again the fix for 1 and 3.

For point 2: It wasnt supposed to be visible the bulk replace icon anymore. Now we can replace one question at a time. Knowing this, as a user, do you think that replacing one question with multiple quesitons is confusing? This is something we are still evaluating.

Points 4 and 5 are things that we can evaluate from the design perspective cc: @marcellamaki @jtamiace.

Just fyi: The reason why we cant save the selected resource in point 5 is because when we uncheck the manual selection mode, we set the number of questions to select to the default value: 10. And if the current resource doesnt have 10 available questions, then it will wait for another resource to be selected until we have 10 available questions in the pool. But we dont communicate this in any way, and I think for the question replacement workflow is a little more confusing because the users never see the questions number input because we dont open the modal in the settings page.

pcenov commented:
Thanks @AlexVelezLl - I confirm that 1 is fixed now.
3 is partially fixed because there is still no snackbar message informing the user that the replacement was successful.

I was also able to identify the following issues:
6. For some reason the "Search" button is not working so we can't test the replacement of a question found by searching for it.
7. When in manual selection mode the link "N questions selected" brings the user to a page where the user can essentially do the same as on the previous page with the addition of the delete icon - so this should probably also be further discussed:


https://github.com/user-attachments/assets/3a962013-ff04-4f67-8b4b-4f4288a06a2d

AlexVelezLl committed:
Refactor base ground for multiple replacement questions and max selection restriction

AlexVelezLl committed:
Add button for bulk questions replacement

AlexVelezLl committed:
Add restrictions for multiple questions replacements

AlexVelezLl committed:
Unify default title managemen, and set proper title for replacement questions

AlexVelezLl committed:
Add snackbar for saved changes

AlexVelezLl committed:
Disable replace button if no selection

AlexVelezLl commented:
Hi @pcenov! Now I have fixed all the poings but 6 and 7. ~6 is still in development~ Not anymore :) you can now click on the search button, and 7 is the expected behaviour for now.

Point 2 has been solved by setting a the max amount of questions to select equals to the number of questions being replaced, and we have added again the "replace" button in the bulk toolbar.

For point 5, now the user dont have a way to switch to random selection, so this shouldnt happen anymore in questions replacement :)

AlexVelezLl committed:
Adapt manual questions selection and replace questions to quiz search

AlexVelezLl committed:
Fix possible null children from topic tree resources

pcenov reviewed (approved):
Thank you so much @AlexVelezLl - this is a great update!

marcellamaki commented on kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/sidePanels/QuizResourceSelection/index.vue:
```diff
@@ -488,14 +488,16 @@
-          tooManyQuestions.value ||
-          settings.value.questionCount.value > settings.value.maxQuestions
-        );
+        const disabledConditions = [
```
this is not blocking, and is a bit of a higher level reflection, but I think many of the changes in the quiz workflow added over the last few months make perfect sense within the context of their own PR, and also as we have gone along adding more conditions and options building out these features, these files are starting to get a bit unwieldy. Especially with some of the underlying complexity from the 0.17 work. 

I'd like us to think about some small, incremental refactors that we can do over the course of our patch releases of 0.18. If you have ideas in mind already, Alex, let's start making note of them :)

marcellamaki reviewed (approved):
Thank you @AlexVelezLl - this looks good. My only comment was a meta-reflection about some of the refactoring work I'd like us to think about (including with @nucleogenesis and @AllanOXDi as it's come up in conversations of PRs there too) and manage in small "bites" over the course of our 0.18 planned patches. 

I'll plan to have some working meetings not probably in the next few weeks, but maybe once 0.18 is ready for translation, and we can do some planning about how to manage the things that we thought "let's come back to this" :)

AlexVelezLl commented:
Sounds good @marcellamaki. I think we can break down this further, and replace some parts of this components with what we have in the `useResourceSelection` composable, perhaps extract all this logic to a `useQuizResourceSelection` composable?, and we can also probably extract the quiz selection settings logic to another composable :). Looking forward to it!
